### PR TITLE
Feature/cards managing

### PR DESCRIPTION
**Pull request by AEugene**

### CODE REVIEW

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 1]
```diff
@@ -186,21 +133,22 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-                    card.frame = cardParameters.frame
-                }
-                
-            }
+        for (index, card) in visibleCards.enumerate() {
+            layoutCard(card, AtIndex: UInt(index))
+        }
+    }
+    
+    private func layoutCard(card: DraggableCardView, AtIndex index: UInt) {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: atIndex

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 2]
```diff
@@ -235,9 +183,8 @@ public class KolodaView: UIView, DraggableCardDelegate {
         return (finalFrame, scale)
     }
     
-    private func moveOtherCardsWithFinishPercent(percent: CGFloat) {
+    internal func moveOtherCardsWithFinishPercent(percent: CGFloat) {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: part, fraction. 'Percentage' is better for Double type

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 3]
```diff
@@ -276,50 +223,18 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-        let firstCardAppearAnimation = POPBasicAnimation(propertyNamed: kPOPViewAlpha)
-        
-        firstCardAppearAnimation.toValue = NSNumber(float: Float(revertCardAnimationToValue))
-        firstCardAppearAnimation.fromValue =  NSNumber(float: Float(revertCardAnimationFromValue))
-        firstCardAppearAnimation.duration = revertCardAnimationDuration
-        firstCardAppearAnimation.completionBlock = {
-            (_, _) in
+    public func applyAppearAnimationIfNeeded() {
+        if let shouldApply = delegate?.koloda(kolodaShouldApplyAppearAnimation: self) where shouldApply == true {
+            self.alpha = 0
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: extraneous self.

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 4]
```diff
@@ -276,50 +223,18 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-        
-        let firstCardAppearAnimation = POPBasicAnimation(propertyNamed: kPOPViewAlpha)
-        
-        firstCardAppearAnimation.toValue = NSNumber(float: Float(revertCardAnimationToValue))
-        firstCardAppearAnimation.fromValue =  NSNumber(float: Float(revertCardAnimationFromValue))
-        firstCardAppearAnimation.duration = revertCardAnimationDuration
-        firstCardAppearAnimation.completionBlock = {
-            (_, _) in
+    public func applyAppearAnimationIfNeeded() {
+        if let shouldApply = delegate?.koloda(kolodaShouldApplyAppearAnimation: self) where shouldApply == true {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: func applyAppearAnimationIfNeeded() {
   if ... { applyAppearAnimation }
}

func applyAppearAnimation() {
}

similarly to UIView design

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 5]
```diff
@@ -276,50 +223,18 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+            self.alpha = 0
+            userInteractionEnabled = false
+            animating = true
             
-            self.animating = false
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: redundant self

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 6]
```diff
@@ -276,50 +223,18 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-            complete?()
+            animator.animateAppearance { [weak self] _ in
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: If this is a trailing 'completion' closure, it's not clear from the context. Could be animateAppearanceWithCompletion {}
At the moment looks like the closure defines the animation itself

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 7]
```diff
@@ -339,24 +254,19 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-                options: .CurveLinear,
-                animations: {
-                    self.moveOtherCardsWithFinishPercent(0)
-                },
-                completion: {
-                    _ in
-                    self.animating = false
+            animating = true
+            animator.resetBackgroundCards { [weak self] _ in
+                if let _self = self {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: guard let _self = self else { return }

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 8]
```diff
@@ -398,101 +303,79 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-                    
-                    frameAnimation.property = POPAnimatableProperty.propertyWithName(kPOPViewFrame) as! POPAnimatableProperty
-                    scaleAnimation.property = POPAnimatableProperty.propertyWithName(kPOPLayerScaleXY) as! POPAnimatableProperty
-                } else {
-                    frameAnimation = POPBasicAnimation(propertyNamed: kPOPViewFrame)
-                    (frameAnimation as! POPBasicAnimation).duration = backgroundCardFrameAnimationDuration
-                    scaleAnimation = POPBasicAnimation(propertyNamed: kPOPLayerScaleXY)
-                    (scaleAnimation as! POPBasicAnimation).duration = backgroundCardFrameAnimationDuration
+            animateCardsAfterLoading { [weak self] in
+                if let _self = self {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: guard

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 9]
```diff
@@ -398,101 +303,79 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+                    _self.visibleCards.last?.hidden = false
+                    _self.animating = false
+                    _self.delegate?.koloda(_self, didSwipedCardAtIndex: UInt(_self.currentCardNumber - 1), inDirection: direction)
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: didSwipe

Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: self.currentCardNumber - 1
no out of range check?

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 10]
```diff
@@ -398,101 +303,79 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-            delegate?.koloda(self, didSwipedCardAtIndex: UInt(currentCardNumber - 1), inDirection: direction)
             animating = false
+            delegate?.koloda(self, didSwipedCardAtIndex: UInt(self.currentCardNumber - 1), inDirection: direction)
             delegate?.koloda(kolodaDidRunOutOfCards: self)
         }
-        
+    }
+    
+    private func loadNextCard() {
+        if let dataSource = dataSource {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: guard let dataSource = dataSource else { return }

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 11]
```diff
@@ -398,101 +303,79 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+            let cardParameters = backgroundCardParametersForFrame(frameForCardAtIndex(UInt(visibleCards.count)))
+            let lastCard = createCardAtIndex(UInt(currentCardNumber + countOfVisibleCards - 1), frame: cardParameters.frame)
+            
+            let scale = cardParameters.scale
+            lastCard.layer.transform = CATransform3DScale(CATransform3DIdentity, scale.width, scale.height, 1.0)
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: 1 instead of 1.0 is also fine

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 12]
```diff
@@ -505,136 +388,115 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
                 addSubview(firstCardView)
                 visibleCards.insert(firstCardView, atIndex: 0)
                 
-                firstCardView.frame = frameForTopCard()
                 
-                applyRevertAnimation(firstCardView, complete: {
-                    self.delegate?.koloda(self, didShowCardAtIndex: UInt(self.currentCardNumber))
+                animating = true
+                animator.applyRevertAnimation(firstCardView, completion: { [weak self] _ in
+                    if let _self = self {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: guard

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 13]
```diff
@@ -505,136 +388,115 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+                        _self.animating = false
+                        _self.delegate?.koloda(_self, didShowCardAtIndex: UInt(_self.currentCardNumber))
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: _self.currentCardNumber - is it count or index?

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 14]
```diff
@@ -505,136 +388,115 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
-                let currentCard = visibleCards[index]
-                
-                currentCard.configure(currentCardContentView, overlayView: overlayView)
+                let actualIndex = UInt(currentCardNumber + index)
+                configureCard(card, atIndex: actualIndex)
             }
         }
     }
     
+    private func calculateMissingCardsCount() -> Int {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: private func missingCardsCount() -> Int

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 15]
```diff
@@ -651,4 +513,114 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+                completion: { _ in
+                    self.removeCards(cards)
+                }
+            )
+        } else {
+            self.removeCards(cards)
+        }
+    }
+    
+    public func insertCardAtIndexSet(indexRange: Range<Int>, animated: Bool = true) {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: insertCardsInRange

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 16]
```diff
@@ -651,4 +513,114 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+        if let dataSource = dataSource {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: guard let dataSource

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 17]
```diff
@@ -651,4 +513,114 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+            let currentItemsCount = countOfCards
+            let visibleIndexes = [Int](indexRange).filter { $0 >= currentCardNumber && $0 < currentCardNumber + countOfVisibleCards }
+            let insertedCards = insertVisibleCardsWithIndexes(visibleIndexes.sort())
+            removeCards(visibleCards.dropFirst(countOfVisibleCards).map { $0 }, animated: animated)
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: what's the purpose of this map? map { $0 }

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 18]
```diff
@@ -651,4 +513,114 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+            assert(
+                currentItemsCount + indexRange.count == countOfCards,
+                "Cards count after update is not equal to data source count"
+            )
+        }
+    }
+    
+    // MARK: Cards managing - Deletion
+    
+    private func proceedDeletion(range: Range<Int>) {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: proceedDeletionInRange

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 19]
```diff
@@ -651,4 +513,114 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+        deletionIndexes.sort { $0 > $1 }.forEach { deletionIndex in
+            let visibleCardIndex = deletionIndex - currentCardNumber
+            let card = visibleCards[visibleCardIndex]
+            card.delegate = nil
+            card.swipe(.Right)
+            visibleCards.removeAtIndex(visibleCardIndex)
+        }
+    }
+    
+    public func removeCardAtIndexRange(indexRange: Range<Int>, animated: Bool) {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: inRange

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 20]
```diff
@@ -651,4 +513,114 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+        if let dataSource = dataSource {
+            animating = true
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: guard

File: Pod/Classes/KolodaView/KolodaView.swift [hunk 21]
```diff
@@ -651,4 +513,114 @@ public class KolodaView: UIView, DraggableCardDelegate {
...
+            let currentItemsCount = countOfCards
+            let visibleIndexes = [Int](indexRange).filter { $0 >= currentCardNumber && $0 < currentCardNumber + countOfVisibleCards }
+            if !visibleIndexes.isEmpty {
+                proceedDeletion(visibleIndexes[0]...visibleIndexes[visibleIndexes.count - 1])
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaView.swift]: proceedDeletion(visibleIndexes[0]...visibleIndexes[visibleIndexes.count - 1])
could be
  proceedDeletion(visibleIndexes[0]..<visibleIndexes[visibleIndexes.count])

File: Pod/Classes/KolodaView/KolodaViewAnimatior.swift [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+        kolodaAppearAlphaAnimation.beginTime = CACurrentMediaTime() + cardSwipeActionAnimationDuration
+        kolodaAppearAlphaAnimation.fromValue = NSNumber(float: 0.0)
+        kolodaAppearAlphaAnimation.toValue = NSNumber(float: 1.0)
+        kolodaAppearAlphaAnimation.duration = 0.8
+        
+        koloda?.pop_addAnimation(kolodaAppearAlphaAnimation, forKey: "kolodaAppearScaleAnimation")
+        koloda?.layer.pop_addAnimation(kolodaAppearScaleAnimation, forKey: "kolodaAppearAlphaAnimation")
+    }
+    
+    public func applyRevertAnimation(card: DraggableCardView, completion: AnimationCompletionBlock = nil) {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaViewAnimatior.swift]: applyReverseAnimation (revert is a verb)

File: Pod/Classes/KolodaView/KolodaViewAnimatior.swift [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+            animations: {
+                cards.forEach { $0.alpha = 0.0 }
+            },
+            completion: { finished in
+                completion?(finished)
+            }
+        )
+    }
+    
+    internal func resetBackgroundCards(completion: AnimationCompletionBlock = nil) {
```
Comment (mmrmmlrr) [inline Pod/Classes/KolodaView/KolodaViewAnimatior.swift]: resetBackgroundCardsWithCompletion

**Comment by AEugene** [comment]
Closes #122 , closes #84 , closes #129

**Comment by AEugene** [comment]
Closes #128

**Pull request merged**